### PR TITLE
[sqlalchemy-oso] Ensure we use true filter where possible.

### PR DESCRIPTION
**Pull request by samscott89**
PR checklist:

- [x] Added changelog entry.

### CODE REVIEW

File: languages/python/sqlalchemy-oso/tests/test_sqlalchemy.py [hunk 1]
```diff
@@ -290,3 +290,16 @@ def test_regular_session(engine, oso, fixture_data):
...
+def test_unconditional_policy_has_no_filter(engine, oso, fixture_data):
+    oso.load_str('allow("user", "read", post: Post) if post.id = 1; allow(_, _, _);')
+    session = AuthorizedSession(oso, user="user", action="read", bind=engine)
+
+    query = session.query(Post)
+
+    assert str(query) == (
+        "SELECT posts.id AS posts_id, posts.contents AS posts_contents, posts.title AS posts_title, "
+        + "posts.access_level AS posts_access_level, posts.created_by_id AS posts_created_by_id, "
+        + "posts.needs_moderation AS posts_needs_moderation \nFROM posts \nWHERE 1 = 1"
```
Comment (samscott89) [inline languages/python/sqlalchemy-oso/tests/test_sqlalchemy.py:304]: This previously had a filter `post.id = ?` _instead_ of a true filter.

File: languages/python/sqlalchemy-oso/tests/test_post_relationship.py [hunk 2]
```diff
@@ -644,9 +644,9 @@ def test_in_with_constraints_but_no_matching_objects(
         + " posts.access_level AS posts_access_level,"
         + " posts.created_by_id AS posts_created_by_id, posts.needs_moderation AS posts_needs_moderation"
         + " \nFROM posts"
-        + " \nWHERE (EXISTS (SELECT 1"
+        + " \nWHERE EXISTS (SELECT 1"
```
Comment (samscott89) [inline languages/python/sqlalchemy-oso/tests/test_post_relationship.py:647]: No idea why. SQLalchemy version thingy?

File: languages/python/sqlalchemy-oso/sqlalchemy_oso/partial.py [hunk 3]
```diff
@@ -41,6 +42,13 @@ def flip_op(operator):
     return flips[operator]
 
 
+def add_filter(current, new):
```
Comment (leina05) [inline languages/python/sqlalchemy-oso/sqlalchemy_oso/partial.py:45]:
```suggestion
def and_filter(current, new):
```

File: languages/python/sqlalchemy-oso/tests/test_post_relationship.py (hunk 2 above)
Comment (leina05) [reply languages/python/sqlalchemy-oso/tests/test_post_relationship.py:647]: Was this failing without your change? Or after the change?

File: languages/python/sqlalchemy-oso/tests/test_sqlalchemy.py (hunk 1 above)
Comment (leina05) [reply languages/python/sqlalchemy-oso/tests/test_sqlalchemy.py:304]: Looks good, we should probably add this test to the django adapter?

**Review by leina05 — APPROVED** [review]
Just documenting for myself/others: the bug here was that we were converting unconditional expressions to empty `AND` clauses, which when `OR`'d with other filters do nothing, when the desired behavior is to turn the filter into a `TRUE`/`WHERE 1 = 1` so that no conditions are applied.

File: languages/python/sqlalchemy-oso/tests/test_sqlalchemy.py (hunk 1 above)
Comment (samscott89) [reply languages/python/sqlalchemy-oso/tests/test_sqlalchemy.py:304]: Good idea.

File: languages/python/sqlalchemy-oso/tests/test_post_relationship.py (hunk 2 above)
Comment (samscott89) [reply languages/python/sqlalchemy-oso/tests/test_post_relationship.py:647]: I can check. I suspect there's an additional `and` that was stripped away from the query filter. e.g. `and(and(...))` vs `and(...)`

**Comment by samscott89** [comment]
> 
> 
> Just documenting for myself/others: the bug here was that we were converting unconditional expressions to empty `AND` clauses, which when `OR`'d with other filters do nothing, when the desired behavior is to turn the filter into a `TRUE`/`WHERE 1 = 1` so that no conditions are applied.

Good summary! We had the same problem in the django adapter, which is why we have the `TRUE_FILTER` throughout.

File: languages/python/sqlalchemy-oso/tests/test_sqlalchemy.py (hunk 1 above)
Comment (samscott89) [reply languages/python/sqlalchemy-oso/tests/test_sqlalchemy.py:304]: Added in https://github.com/osohq/oso/pull/861/commits/03715a5e520289fdd0fc344286433a44f981f29a

File: languages/python/sqlalchemy-oso/sqlalchemy_oso/partial.py (hunk 3 above)
Comment (samscott89) [reply languages/python/sqlalchemy-oso/sqlalchemy_oso/partial.py]: Added in https://github.com/osohq/oso/pull/861/commits/03715a5e520289fdd0fc344286433a44f981f29a

**Pull request merged**